src/rust/ide/json-rpc/src/handler.rs

```diff
@@ -276,7 +290,14 @@ impl<Notification> Handler<Notification> {
             // If message cannot be send, future ret must be cancelled.
             self.remove_ongoing_request(id);
         }
-        ret
+
+        let timeout = self.timeout();
+        future::select(ret, sleep(timeout).boxed()).map(move |either|
+            match either {
+                future::Either::Left ((x, _)) => x,
+                future::Either::Right((_, _)) => Err(RpcError::TimeoutError{millis:timeout.as_millis()}),
+            }
```

Comment:
There has to be a function for that. It is too popular operation to require pattern matching

Reply:
I am not aware of such function.